rust/benchmark-datasets/src/types.rs

```diff
@@ -0,0 +1,235 @@
+use anyhow::Result;
+use rand::seq::SliceRandom;
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, future::Future, path::PathBuf};
+use tantivy::{
+    collector::TopDocs,
+    doc,
+    query::QueryParser,
+    schema::{Schema, STORED, TEXT},
+    Index, ReloadPolicy,
+};
+use tokio::io::AsyncWriteExt;
+use tokio_stream::{Stream, StreamExt};
+
+use crate::util::{get_dir_for_persistent_dataset_files, get_or_populate_cached_dataset_file};
+
+trait TantivyIndexable {
+    fn build_tantivy_index(&mut self) -> impl Future<Output = Result<(Index, Schema)>>;
+}
+
+impl<T: Stream<Item = Document> + Unpin> TantivyIndexable for T {
+    async fn build_tantivy_index(&mut self) -> Result<(Index, Schema)> {
+        let mut schema_builder = Schema::builder();
+        schema_builder.add_text_field("content", TEXT | STORED);
+        let schema = schema_builder.build();
+
+        let index = Index::create_in_ram(schema.clone());
+        let mut index_writer = index.writer(50_000_000)?;
+        let content = schema.get_field("content")?;
+
+        while let Some(document) = self.next().await {
+            index_writer.add_document(doc!(content => document.content))?;
+        }
+        index_writer.commit()?;
+
+        Ok((index, schema))
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct FrozenQuerySubset {
+    queries: Vec<String>,
+    queries_from_dataset: String,
+    dataset_queries_tested_against: String,
+    min_results_per_query: usize,
+}
+
+#[derive(Debug, Clone)]
+pub struct Document {
```

Comment:
maybe we want to call this "Record" and "Content" to "Document" - in the future if we want to add embeddings and make this usable for more than FTS it will be useful.